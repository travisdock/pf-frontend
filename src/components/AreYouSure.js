import React from 'react';
import Popup from "reactjs-popup";

class AreYouSure extends React.Component {
    render() {
        return (
            <Popup
            open={this.props.open === "ask"}
            closeOnDocumentClick
            onClose={this.props.closeModal}
            onOpen={this.props.onOpen}
            >
            <div className="table-popup">
                <p>{this.props.message}</p>
                <button className="delete" onClick={this.props.handleDelete} >Delete</button>
                <button className="close" onClick={this.props.closeModal} >Close</button>
            </div>
            </Popup>
        )
    }
};

export default AreYouSure;