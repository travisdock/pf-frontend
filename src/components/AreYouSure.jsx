import React from 'react';
import Popup from "reactjs-popup";

class AreYouSure extends React.Component {
    render() {
        return (
            <Popup
            open={this.props.open === "ask"}
            closeOnDocumentClick
            onClose={this.props.closeModal}
            onOpen={this.props.askIfSure}
            >
            <div className="table-popup">
                <p>Are You Sure?</p>
                <button onClick={this.props.handleDelete} >Delete</button>
                <button onClick={this.props.closeModal} >Close</button>
            </div>
            </Popup>
        )
    }
};

export default AreYouSure;