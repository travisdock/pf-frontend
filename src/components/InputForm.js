import React from 'react';

import { connect } from 'react-redux';

import { updateUser } from '../actions/UserActions';
import {
  handleChange, handleSubmit,
  toggleCategory, toggleIncome,
  toggleUntracked, evaluateAmount
} from './helpers/inputFormHelpers'

class InputForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      form: {
        category_name: '',
        date: new Date().toISOString({timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone}).substr(0, 10),
        amount: '',
        notes: '',
        income: false,
        untracked: false
      },
      new_category: false
    };
    // Imported functions
    this.handleChange = handleChange.bind(this);
    this.handleSubmit = handleSubmit.bind(this);
    this.toggleCategory = toggleCategory.bind(this);
    this.toggleIncome = toggleIncome.bind(this);
    this.toggleUntracked = toggleUntracked.bind(this);
    this.evaluateAmount = evaluateAmount.bind(this);
  }

  render() {
    const newCategory = (<div className="ui field">
      <input
        name="category_name"
        value={this.state.form.category_name}
        placeholder='Category Name'
        onChange={this.handleChange}
      />
      <div className="checkbox_wrapper">
        <label className="checkboxes">
        <input
          name="income"
          type="checkbox"
          checked={this.state.form.income}
          onChange={this.toggleIncome}
        />income</label>

        <label className="checkboxes">
        <input
          name="untracked"
          type="checkbox"
          checked={this.state.form.untracked}
          onChange={this.toggleUntracked}
        />untracked</label>
      </div>
    </div>)
    const selectCategory = (<div className="ui field">
      <select
        name="category_name"
        value={this.state.form.category_name}
        onChange={this.handleChange}
        >
        <option value="" disabled>Select Category</option>
        {!!this.props.current_user.categories ? this.props.current_user.categories.map(cat => <option value={cat.name} key={cat.id}>{cat.name}</option>) : null}
      </select>
    </div>)
    return (
      <div className="input_form">
        <form onSubmit={this.handleSubmit}>
          <label>
            <input
              name="newcategory"
              type="checkbox"
              onChange={this.toggleCategory}
              checked={this.state.new_category}
              />
            New Category?
          </label>
          {(this.state.new_category ? newCategory : selectCategory)}
            <input
              name="date"
              type="date"
              value={this.state.form.date}
              onChange={this.handleChange}
            />
            <input
              name="amount"
              placeholder="0.00"
              type="text"
              value={this.state.form.amount}
              onChange={this.handleChange}
              onBlur={this.evaluateAmount}
            />
            <textarea
              name="notes"
              placeholder="Entry details..."
              rows="5"
              cols="55"
              value={this.state.form.notes}
              onChange={this.handleChange}
            />
          <button type="submit" className="button">
            Submit
          </button>
        </form>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  current_user: state.current_user
});

export default connect(mapStateToProps, { updateUser })(InputForm);
