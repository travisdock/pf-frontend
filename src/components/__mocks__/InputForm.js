import React from 'react';

export default class InputForm extends React.Component {
  render() {
    return (<div className="content"></div>)
  }
}
