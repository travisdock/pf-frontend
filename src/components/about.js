import React from 'react';

const About = () => (
  <div className="about">
    <h1>Sum.</h1>
    <p>Some guy once said something like "What gets measured, gets managed". For people having trouble managing their personal finances, simply measuring their income and expenses is the first step. Often just the act of knowing where your money is going brings about the motivation to make changes. Sum. is a very simple app that gives the user a lot of freedom in simply measuring. Create any number of custom categories for income and expenses. Then watch where the money goes. Once you actually see how your expenses add up it will almost be harder not to make any changes.</p>
  </div>
)

export default About;
