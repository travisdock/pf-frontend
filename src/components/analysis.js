import React from 'react';
import { connect } from 'react-redux';

import Chart from './chart';

class Analysis extends React.Component {
  state = {
    userEntries: []
  }

  componentDidMount() {
    const url = `http://localhost:3001/api/v1/month_category/${this.props.current_user.user_id}`
    fetch(url)
      .then(resp => resp.json())
      .then(resp => this.setState({userData: resp}))
  }
  render(){

    return(
      <div>{this.state.userData ?
        <div><Chart /></div>
        :
        <div>Loading...</div>
      }</div>
    )
  }
}

const mapStateToProps = state => ({
  current_user: state.current_user
});

export default connect(mapStateToProps)(Analysis);
