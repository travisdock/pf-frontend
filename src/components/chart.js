import React from 'react';
import {bb} from "billboard.js";
import { connect } from 'react-redux';

class Chart extends React.Component {

  state = {
    charts: {},
    currentChart: 0,
    stats: {},
    extraMonthInfo: {},
    load: false
  }

  componentDidMount() {
    const id = this.props.current_user.user_id

    fetch(`${process.env.REACT_APP_API}/api/v1/charts/${id}`)
      .then(resp => resp.json())
      .then(resp => {if (resp.error) {
        this.setState({error: resp.error})
      } else {
        const titles = Object.keys(resp.charts)
        const lastMonth = titles[titles.length - 1]
        this.setState({
          charts: resp.charts,
          stats: resp.stats,
          load: true,
          currentChart: resp.charts[lastMonth], currentTitle: lastMonth
        }, this.renderPieChart)
      }}
        )
  }

  selectChangePieChart = (e) => {
    const title = e.target.value
    this.setState( () => {
      return {
        currentChart: this.state.charts[title],
        currentTitle: title
      }
    }, this.renderPieChart)
  }

  renderPieChart() {
    const { currentChart, currentTitle } = this.state
    // const { currentTitle } = Object.keys(currentChart)[0]
    // debugger;
    bb.generate({
      data: {
        columns: currentChart,
        type: "pie",
        // onclick: function(d, element) { debugger }
      },
      title: {
        text: currentTitle,
        position: "top-center",
        padding: {
            top: 10,
           right: 10,
            bottom: 10,
            left: 10
        },
      },
      legend: {
        show: true,
        // contents: {
        //     bindto: "#legend",   // <ul id='legend'></ul>
        //     template: "<li style='background-color:#1f77b4'>data1</li>"
        //     or using function
        //     template: function(id, color, data) {
        //          // if you want omit some legend, return falsy value
        //          if (title !== "data1") {
        //               return "<li style='background-color:"+ color +">"+ title +"</li>";
        //          }
        //     }
        // },
        position: "bottom",  // bottom, right, inset
        equally: false,
        padding: 10,
        item: {
            // onclick: function(id) { debugger },
            // onover: function(id) { debugger },
            // onout: function(id) { debugger },
    
            // set tile's size
            tile: {
                width: 20,
                height: 15
            }
        }
      },
      bindto: "#chart"
    });
  }

  render() {
    return (
      <div className="content">
        {this.state.load ?
            <div className="chart-content">
              <select
                name="chart"
                value={this.state.currentTitle} //current chart month
                onChange={this.selectChangePieChart}
                >
                {Object.keys(this.state.charts).map(title => <option value={title} key={title} > {title} </option>) }
              </select>
              <div id="chart" />
            </div> :
            this.state.error ? <div>{this.state.error}</div> : <div>Loading...</div>
        }
      </div>
    )
  }
}
const mapStateToProps = state => ({
  current_user: state.current_user
});

export default connect(mapStateToProps)(Chart)



// This is the profit loss chart. I got rid of it because it was ugly but I may want to remake it one day so here is the code.
// renderPLChart() {
//   const { currentChart } = this.state
//   bb.generate({
//     data: {
//       x: "x",
//       columns: currentChart,
//     type: "area-spline"
//     },
//     axis: {
//       x: {
//         type: "timeseries",
//         tick: {
//           format: "%b"
//         }
//       }
//     },
//     bindto: "#chart"
//   });
// }