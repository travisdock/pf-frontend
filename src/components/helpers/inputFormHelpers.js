const math = require('mathjs-expression-parser')
export function handleChange(e) {
    const target = e.target;
    const value = target.value;
    const name = target.name;

    this.setState((prevState) => {
      return {
        ...prevState,
        form: {
          ...prevState.form,
          [name]: value
        }
      }
    });
  };

  export function handleSubmit(e) {
    e.preventDefault();
    const url = `${process.env.REACT_APP_API}/api/v1/entries`
    const token = localStorage.getItem('jwt')
    const formData = {
      ...this.state.form,
      user_id: this.props.current_user.user_id
    }
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Authorization': token
      },
      body: JSON.stringify(formData)
    }
    fetch(url, options)
      .then(res => res.json())
      .then(resp => {
        if (resp.errors) {
          alert(resp.errors)
        } else if (resp.error) {
          alert(resp.error)
        } else {
          if (resp.id) {
            this.props.updateUser(resp)
          }
          alert("Success!")
          this.setState({
            form: {
              category_name: '',
              date: new Date().toISOString().substr(0, 10),
              amount: '',
              notes: '',
              income: false,
              untracked: false
            },
            new_category: false
          })
        }
      })
  }

  export function toggleCategory() {
    this.setState((prevState) => {
      return {
        ...prevState,
        form: {
          ...prevState.form,
          category_name: '',
          income: false,
          untracked: false
        },
        new_category: !prevState.new_category
      }
    })
  }

  export function toggleIncome() {
    this.setState((prevState) => {
      return {
        ...prevState,
        form: {
          ...prevState.form,
          income: !prevState.form.income,
          untracked: false
        }
      }
    })
  }
  export function toggleUntracked() {
    this.setState((prevState) => {
      return {
        ...prevState,
        form: {
          ...prevState.form,
          untracked: !prevState.form.untracked,
          income: false
        }
      }
    })
  }
  export function evaluateAmount(e) {
      if (e.target.value) {
          try {
              e.target.value = math.eval(e.target.value);
              this.handleChange(e);
          }
          catch(err) {
              alert(`There was an error: ${err.message}`)
          }
        }
  };