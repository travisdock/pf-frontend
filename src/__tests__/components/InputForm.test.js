import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { InputForm } from '../../components/InputForm';
import * as Helpers from '../../components/helpers/inputFormHelpers';

const props = {
    current_user: {
        user_id: 1,
        categories: [
            {
                id: 1,
                name: "categoryOne"
            },
            {
                id: 2,
                name: "categoryTwo"
            }
        ]
    }
};
const noCategoriesUserProps = {
    current_user: {
        user_id: 1
    }
}

describe('snapshots', () => {
    beforeEach(() => {
        let date = new Date('1993')
        global.Date = jest.fn(() => date)
    })
    
    afterAll(() => {
        global.Date = Date
    })

    it('renders an <InputForm /> with select category snapshot', () => {
        const component = renderer.create(<InputForm {...props} />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders an <InputForm /> with no user categories', () => {
        const component = renderer.create(<InputForm {...noCategoriesUserProps} />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders an <InputForm /> with new category snapshot', () => {
        const component = renderer.create(<InputForm {...props} />);
        component.root.instance.setState({new_category: true})
        
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('select Category methods fire appropriately', () => {
    test('handleSubmit is fired on button click', () => {
        const spy = jest.spyOn(Helpers, "handleSubmit");

        const component = shallow(<InputForm {...props} />);

        const form = component.find('form')
        form.simulate('submit', { preventDefault () {} });

        expect(spy).toHaveBeenCalled();
    });
    
    it('calls toggle category on checkbox check', () => {
        const spy = jest.spyOn(Helpers, "toggleCategory");
        const component = shallow(<InputForm {...props} />);
        
        const input = component.find('input[type="checkbox"]').first()
        input.simulate('change', { target: { checked: true } });
        
        expect(spy).toHaveBeenCalled();
    });

    it('calls handleChange on category select change', () => {
        const spy = jest.spyOn(Helpers, "handleChange");
        const component = shallow(<InputForm {...props} />);
        
        const input = component.find('select').first()
        input.simulate('change', {target: { value: 'categoryTwo' } });
        
        expect(spy).toHaveBeenCalled();
    });

    it('calls handleChange on date input change', () => {
        const spy = jest.spyOn(Helpers, "handleChange");
        const component = shallow(<InputForm {...props} />);
        
        const input = component.find('input[type="date"]').first()
        input.simulate('change', {target: { value: '12/22/16' } });
        
        expect(spy).toHaveBeenCalled();
    });

    it('calls handleChange on amount input change', () => {
        const spy = jest.spyOn(Helpers, "handleChange");
        const component = shallow(<InputForm {...props} />);
        
        const input = component.find('input[type="text"]').first()
        input.simulate('change', {target: { value: '122' } });
        
        expect(spy).toHaveBeenCalled();
    });

    it('calls handleChange on notes input change', () => {
        const spy = jest.spyOn(Helpers, "handleChange");
        const component = shallow(<InputForm {...props} />);
        
        const input = component.find('textarea').first()
        input.simulate('change', {target: { value: 'hello there' } });
        
        expect(spy).toHaveBeenCalled();
    });
});

describe('new Category methods fire appropriately', () => {
    it('calls handleChange on category name input change', () => {
        const spy = jest.spyOn(Helpers, "handleChange");
        const component = shallow(<InputForm {...props} />);
        component.setState({new_category: true})
        component.update()
        
        const input = component.find('input[name="category_name"]').first()
        input.simulate('change', {target: { value: 'new one' } });
        
        expect(spy).toHaveBeenCalled();
    });

    it('calls toggleIncome on checkbox check', () => {
        const spy = jest.spyOn(Helpers, "toggleIncome");
        const component = shallow(<InputForm {...props} />);
        component.setState({new_category: true})
        component.update()
        
        const input = component.find('input[name="income"]').first()
        input.simulate('change', {target: { checked: true } });
        
        expect(spy).toHaveBeenCalled();
    });

    it('calls toggleUntracked on checkbox check', () => {
        const spy = jest.spyOn(Helpers, "toggleUntracked");
        const component = shallow(<InputForm {...props} />);
        component.setState({new_category: true})
        component.update()
        
        const input = component.find('input[name="untracked"]').first()
        input.simulate('change', {target: { checked: true } });
        
        expect(spy).toHaveBeenCalled();
    });
});