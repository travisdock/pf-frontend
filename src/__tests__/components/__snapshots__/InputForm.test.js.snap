// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots renders an <InputForm /> with new category snapshot 1`] = `
<div
  className="input_form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      <input
        checked={true}
        name="newcategory"
        onChange={[Function]}
        type="checkbox"
      />
      New Category?
    </label>
    <div
      className="ui field"
    >
      <input
        name="category_name"
        onChange={[Function]}
        placeholder="Category Name"
        value=""
      />
      <div
        className="checkbox_wrapper"
      >
        <label
          className="checkboxes"
        >
          <input
            checked={false}
            name="income"
            onChange={[Function]}
            type="checkbox"
          />
          income
        </label>
        <label
          className="checkboxes"
        >
          <input
            checked={false}
            name="untracked"
            onChange={[Function]}
            type="checkbox"
          />
          untracked
        </label>
      </div>
    </div>
    <input
      name="date"
      onChange={[Function]}
      type="date"
      value="1993-01-01"
    />
    <input
      name="amount"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.00"
      type="text"
      value=""
    />
    <textarea
      cols="55"
      name="notes"
      onChange={[Function]}
      placeholder="Entry details..."
      rows="5"
      value=""
    />
    <button
      className="button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`snapshots renders an <InputForm /> with no user categories 1`] = `
<div
  className="input_form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      <input
        checked={false}
        name="newcategory"
        onChange={[Function]}
        type="checkbox"
      />
      New Category?
    </label>
    <div
      className="ui field"
    >
      <select
        name="category_name"
        onChange={[Function]}
        value=""
      >
        <option
          disabled={true}
          value=""
        >
          Select Category
        </option>
      </select>
    </div>
    <input
      name="date"
      onChange={[Function]}
      type="date"
      value="1993-01-01"
    />
    <input
      name="amount"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.00"
      type="text"
      value=""
    />
    <textarea
      cols="55"
      name="notes"
      onChange={[Function]}
      placeholder="Entry details..."
      rows="5"
      value=""
    />
    <button
      className="button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`snapshots renders an <InputForm /> with select category snapshot 1`] = `
<div
  className="input_form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      <input
        checked={false}
        name="newcategory"
        onChange={[Function]}
        type="checkbox"
      />
      New Category?
    </label>
    <div
      className="ui field"
    >
      <select
        name="category_name"
        onChange={[Function]}
        value=""
      >
        <option
          disabled={true}
          value=""
        >
          Select Category
        </option>
        <option
          value="categoryOne"
        >
          categoryOne
        </option>
        <option
          value="categoryTwo"
        >
          categoryTwo
        </option>
      </select>
    </div>
    <input
      name="date"
      onChange={[Function]}
      type="date"
      value="1993-01-01"
    />
    <input
      name="amount"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.00"
      type="text"
      value=""
    />
    <textarea
      cols="55"
      name="notes"
      onChange={[Function]}
      placeholder="Entry details..."
      rows="5"
      value=""
    />
    <button
      className="button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
